### ISSUE TITLE
Create neuronal network data Tibetan

### ISSUE THREAD

**Issue by ayya-vimala** [ayya-vimala/pali-networks#10]
* Run the neuronal network for Tibetan data only.
* Create csv files
* check csv files and feedback

**Comment by ayya-vimala** [comment]
"0.0#TD4093.1E.TXT.r#794426#794433#TD4093.1E.TXT.r_4628;TD4093.1E.TXT.r_4630;TD4093.1E.TXT.r_4629;TD4093.1E.TXT.r_4631# de ni de dang mtshungs pa'o // / +gang rang gi las byed pa de ni brten pa dang bcas pa'o zhes bshad par 'gyur ro // / 

translates to:
probability: 0.0
textname: TD4093.1E.TXT.r
beginning position of quote: 794426
end position of quote: 794433
segment name: TD4093.1E.TXT.r_4628;TD4093.1E.TXT.r_4630;TD4093.1E.TXT.r_4629;TD4093.1E.TXT.r_4631
text: de ni de dang mtshungs pa'o // / +gang rang gi las byed pa de ni brten pa dang bcas pa'o zhes bshad par 'gyur ro // / 

There are however 4 segment numbers and only 2 segments in the text. How does that work and why are the segment numbers not in consecutive order?

**Comment by sebastian-nehrdich** [comment]
I have just uploaded new versions of the csv files. now the issue with the "// /" is fixed. Also the segment numbers should now match the segments.
I think you mentioned in a mail, but maybe it is better to keep discussing it here: I agree that replacing "/" with "|" in the display is a good idea and looks much nicer. if it doesn't bother you I would like to continue using "/" in the csv files because "|" is annoying when writing regex (it always needs to be escaped and this turns writing regex into a game of mikado). I decided to change the segment delimiter from " / " to "+" because in many texts, the numbers of "/" actually matters. So a single / might indicate a half stop like a , in English, and a // is more like a full-stop.

**Comment by ayya-vimala** [comment]
I'm happy just to replace the danda in the frontend and fine to have / in the csv files.

**Comment by ayya-vimala** [comment]
I have not seen your new files yet, but this is a weird result I found in the old ones. I will check in the new ones if it is still there once I get them:

In TD4071.csv for segments 1190 to 1194:

> "406847","TD4071E.TXT.r_1190"," ci ste gzhan ji ltar bcom ldan 'das kyis sred pa 'byung  ba'i tshor ba khyad par du mdzad pa de bzhin du tshor ba la sogs pa 'byung ba'i reg pa yang kha cig tu khyad par du mdzad do snyam du sems su dogs nas / / "
> "406867","TD4071E.TXT.r_1191;TD4071E.TXT.r_1192"," reg pa ni zhes bya ba la sogs pa gsungs so // / +de bas na zhes bya ba la sogs pa ni mjug sdud pa yin no // / "
> "406876","TD4071E.TXT.r_1193;TD4071E.TXT.r_1194"," gzhan dag  na re zhes bya ba la sogs pas ni gzhan dag gzhan du lan 'debs pa yin te / / +'di skad du gal te reg pa gang du yang khyad par ma mdzad du zin kyang / / "

So these segments seem just fine. But then there are some parallels mentioned in TD4063E (why not the other way around btw):

> "2406866","TD4063E.TXT.r_5;TD4063E.TXT.r_6"," 'jam dpal gzhon nur gyur pa la phyag 'tshal lo // / +bcom ldan 'das kyis chos thams cad bdag med do zhes gsungs pa ni chos thams cad    ces bya ba gang yin pa dang  bdag med pa de gang la brjod ce na / / ","0.0#TD4071E.TXT.r#406861#406870#TD4071E.TXT.r_1190;TD4071E.TXT.r_1191;TD4071E.TXT.r_1192# ci ste gzhan ji ltar bcom ldan 'das kyis sred pa 'byung  ba'i tshor ba khyad par du mdzad pa de bzhin du tshor ba la sogs pa 'byung ba'i reg pa yang kha cig tu khyad par du mdzad do snyam du sems su dogs nas / / ","0.0#TD4071E.TXT.r#406870#406879#TD4071E.TXT.r_1191;TD4071E.TXT.r_1192;TD4071E.TXT.r_1193;TD4071E.TXT.r_1194# reg pa ni zhes bya ba la sogs pa gsungs so // / +de bas na zhes bya ba la sogs pa ni mjug sdud pa yin no // / "

Just to make it a bit more clear and like I said, this might already have been corrected but just to check:
So in the source file, the segments are:

1190
ci ste gzhan ji ltar bcom ldan 'das kyis sred pa 'byung ba'i tshor ba khyad par du mdzad pa de bzhin du tshor ba la sogs pa 'byung ba'i reg pa yang kha cig tu khyad par du mdzad do snyam du sems su dogs nas / /
1191
reg pa ni zhes bya ba la sogs pa gsungs so // /
1192 
de bas na zhes bya ba la sogs pa ni mjug sdud pa yin no // /
1193
gzhan dag na re zhes bya ba la sogs pas ni gzhan dag gzhan du lan 'debs pa yin te / /
1194
'di skad du gal te reg pa gang du yang khyad par ma mdzad du zin kyang / /

But as parallels they are listed as paralles to:
TD4063E segments 5 and 6
jam dpal gzhon nur gyur pa la phyag 'tshal lo // / 
bcom ldan 'das kyis chos thams cad bdag med do zhes gsungs pa ni chos thams cad ces bya ba gang yin pa dang bdag med pa de gang la brjod ce na / / 

1190-1192
ci ste gzhan ji ltar bcom ldan 'das kyis sred pa 'byung ba'i tshor ba khyad par du mdzad pa de bzhin du tshor ba la sogs pa 'byung ba'i reg pa yang kha cig tu khyad par du mdzad do snyam du sems su dogs nas / / 
1191-1194
reg pa ni zhes bya ba la sogs pa gsungs so // / +de bas na zhes bya ba la sogs pa ni mjug sdud pa yin no // /

**Comment by sebastian-nehrdich** [comment]
Thank you for point this out! I just uploaded the new csv-files (the files with the menustructure etc. have been updated accordingly). I will check now if the issues you mentioned in the last post also occur here.

**Comment by ayya-vimala** [comment]
This one seems indeed better but will check over the next days if I find any strange occurances. The main problem for the highlighting is that segments that are taken together in the source files and in the parallels are not always the same. So it tends just to highlight a part and not the whole one. Not sure what to do about that yet.

**Comment by sebastian-nehrdich** [comment]
Sorry for asking so stupidly: What do you mean by this? Maybe I can try to find a way to fix things in the data generation. With Tibetan, we can actually be more precise with the matches than with the Sanskrit. But it doesn't yet become very clear in the results. This is why I have implemented it now that the parallel words in the quote segment are in uppercase (I don't know if I discussed this with you before). We could think of changing this, I am also happy to give it a span-class or something like this. I wanted to use a solution with which I do not have to change the word/character count of the segments.

**Comment by sebastian-nehrdich** [comment]
another thing: I have changed the "_" seperator in the Tibetan segment names to ":" in order to have better consistency with the Sanskrit and Tibetan files. The new Tibetan csv files are already online.

**Comment by ayya-vimala** [comment]
What I mean is that for instance in TD4063E:

> "10222","T06TD4063E:5;T06TD4063E:6"," 'jam dpal gzhon nur gyur pa la phyag 'tshal lo //+bcom ldan 'das kyis chos thams cad bdag med do zhes gsungs pa ni chos thams cad    ces bya ba gang yin pa dang  bdag med pa de gang la brjod ce na /",
> "0.015#T06TD4027E#10634#10634#T06TD4027E:6# 'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //",
> "0.024#T06TD4074E#2839245#2839245#T06TD4074E:5;T06TD4074E:6# 'phags pa     // +'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //"

So the first parallels is with TD4027E segment 6. Sofar fine. 
But when I click on that parallel in the list, it loads TD4027E and there it takes all the segments from that sutta to display it. And segment 6 is only one segment of a combination of segments 4-6:

> "10628","T06TD4027E:4;T06TD4027E:5;T06TD4027E:6"," dbus dang mtha' rnam par 'byed pa'i 'grel pa /+bam po dang po /+'jam dpal gzhon nur gyur pa la phyag 'tshal lo //"

So it highlights all segments 4-6.

This is what I mean that parallels do not match up with the combined segments in the sutta itself. 
This could be okay because people are very capable of seeing which part is matching up. I don't think you need to make that in uppercase.
BTW Is it okay to have the whole text in uppercase? I thought that is normally done with Tibetan but I might be wrong.

Right now it looks like this:
![Screenshot from 2019-06-21 13-59-55](https://user-images.githubusercontent.com/49606945/59921252-64609080-942d-11e9-9f5c-68b60acfb5ec.png)

Will try to upload stuff this evening. I'm on a mobile data plan and I have to be careful what I do in the daytime but evenings I have more space.

**Comment by ayya-vimala** [comment]
BTW. Why does it say:

> "0.024#T06TD4074E#2839245#2839245#T06TD4074E:5;T06TD4074E:6# 'phags pa     // +'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //"

In the above example. Surely as only "jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //" (jam should also be in uppercase) is the parallel, why add the segment number 5: "phags pa     //" to that? Why not just have segment 6 as parallel?

**Comment by ayya-vimala** [comment]
With regards to the Sanskrit files you mention in your README:

> The layout of the Sanskrit matches is the same as in the case of Tibetan, but there are some slight differences in the segments:  
> sthtvbhu:P:3_4
> The structure is: filename:segmenttype:number
> segmenttype can be either P or LG. LG is a line-group, meaning a Kārika. P is paragraph. A new paragraph starts when the paragraph number ends in zero (e.h. 3_0 means paragraph 3 begins).  
> There is also sometimes HTML (bold-tags and italics) in the Sanskrit text to mark certain sections. Not yet sure what to do with this (convert into a css-class?).  

I cannot quite see the value for having segmenttypes in the csv for me. The parallels are just displayed by themselves in a little box so there it is of no value. The texts themselves I can filter for the '0' and start a new paragraph there.

The bold and italics ... not sure. I will do some tests to see what it does with adding an html tag in there. I think it will just display the tag as part of the text. Will have a think about it.

**Comment by ayya-vimala** [comment]
One more thing I noticed about the Sanskrit files: the '+' in between segments does not seem to be there:

> "649261","abhisamu:LG:Abhs_1.1_1;abhisamu:LG:Abhs_1.2_0"," dhīmantī vīkṣīṣīraṃstamanālīḍhaṃ parairiti // / smṛtau cādhāya sūtrārthaṃ dharmacaryā daśātmikā /"

It only appears in the parallels-matches.

**Comment by sebastian-nehrdich** [comment]
> BTW. Why does it say:
> 
> > "0.024#T06TD4074E#2839245#2839245#T06TD4074E:5;T06TD4074E:6# 'phags pa     // +'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //"
> 
> In the above example. Surely as only "jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //" (jam should also be in uppercase) is the parallel, why add the segment number 5: "phags pa //" to that? Why not just have segment 6 as parallel?
The reason for this is: The minimum match length is 8 words (on the side of the algorithm, in order to avoid over-generation of junk). This is why we cannot go for segments smaller than this and "phags pa //" also shows up.

**Comment by sebastian-nehrdich** [comment]
> What I mean is that for instance in TD4063E:
> 
> > "10222","T06TD4063E:5;T06TD4063E:6"," 'jam dpal gzhon nur gyur pa la phyag 'tshal lo //+bcom ldan 'das kyis chos thams cad bdag med do zhes gsungs pa ni chos thams cad    ces bya ba gang yin pa dang  bdag med pa de gang la brjod ce na /",
> > "0.015#T06TD4027E#10634#10634#T06TD4027E:6# 'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //",
> > "0.024#T06TD4074E#2839245#2839245#T06TD4074E:5;T06TD4074E:6# 'phags pa     // +'jam DPAL GZHON NUR GYUR PA LA PHYAG 'TSHAL LO //"
> 
> So the first parallels is with TD4027E segment 6. Sofar fine.
> But when I click on that parallel in the list, it loads TD4027E and there it takes all the segments from that sutta to display it. And segment 6 is only one segment of a combination of segments 4-6:
> 
> > "10628","T06TD4027E:4;T06TD4027E:5;T06TD4027E:6"," dbus dang mtha' rnam par 'byed pa'i 'grel pa /+bam po dang po /+'jam dpal gzhon nur gyur pa la phyag 'tshal lo //"
I get it! I described further below how we might be able to deal with it. 
> 
> So it highlights all segments 4-6.
> 
> This is what I mean that parallels do not match up with the combined segments in the sutta itself.
> This could be okay because people are very capable of seeing which part is matching up. I don't think you need to make that in uppercase.
Yes I agree with that. In the case of Tibetan, segments are short & related parts are easily found. Might be more of an issue with Pāli, where the segments are much bigger...
> BTW Is it okay to have the whole text in uppercase? I thought that is normally done with Tibetan but I might be wrong.
Good question! I am used to read Wylie in lowercase (because there is extended Wylie as well which uses uppercase and lowercase). I have to discuss this with the team. Would of course also be nice to have a display in Tibetan letters. That should not be hard to do, I am sure there is some npm package for that out there. But not urgent. 
> 
> Right now it looks like this:
> ![Screenshot from 2019-06-21 13-59-55](https://user-images.githubusercontent.com/49606945/59921252-64609080-942d-11e9-9f5c-68b60acfb5ec.png)
> 
> Will try to upload stuff this evening. I'm on a mobile data plan and I have to be careful what I do in the daytime but evenings I have more space.
Thank you! It is very clear to me now what you mean. I could also change the way the files are created. At the moment, we get the "combined" segments with multiple IDs and pulled together sentences combined with a '+' in the leftmost column. I could change that and have only single columns in the leftmost column (and nothing pulled together with a '+') and store the quote just at the segment where it is starting and not where the main body of the quote is located (as it is done now: I am approximating the segments that have the highest overlap with the quote, which can be a little bit too early or too late at time, since it is approximate only). Its not a bad idea to change this I think and solves many of the issues. A new problem however will arise: If the segment where the quote is stored is short (for example just two or three words), but the quote is long (for example 10 words), we need to find a way to also highlight the rest in the interface. One way to do that (but we have to do it on the frontend-side): read in the length of the source (leftmost) segment and compare it with the length of the assigned quote. When the length of the assigned quotes exceeds the length of the source quote by x words, we automaticly highlight the next segment as well. This is computationally cheap and somewhat imprecise, but I imagine it can work very well for our problem.

**Comment by sebastian-nehrdich** [comment]
> One more thing I noticed about the Sanskrit files: the '+' in between segments does not seem to be there:
> 
> > "649261","abhisamu:LG:Abhs_1.1_1;abhisamu:LG:Abhs_1.2_0"," dhīmantī vīkṣīṣīraṃstamanālīḍhaṃ parairiti // / smṛtau cādhāya sūtrārthaṃ dharmacaryā daśātmikā /"
> 
> It only appears in the parallels-matches. And the '// /' seems to be back everywhere.

Ouch yes, didn't check it. Will fix that tomorrow. But let's do a test-run of the Tibetan files in the changed layout (as described in my previous posting) first and if that turns out to work well, I will have to change the Sanskrit scripts anyway and I will fix these bugs at that occasion.

**Comment by ayya-vimala** [comment]
> Thank you! It is very clear to me now what you mean. I could change the way the files are created. At the moment, we get the "combined" segments with multiple IDs and pulled together sentences combined with a '+' in the leftmost column

I actually found a solution in meditation. I can deal with that in the frontend by making sure that the segment is represented not with a string but with an array. Then it should be easy. Will work on that.

**Comment by sebastian-nehrdich** [comment]
Yes I see what you mean (and meditation usually brings about the best ideas anyway!). I did something very similar when I was working on the JQUERY frontend.
(below I use the word 'root segment' to describe the segments in the leftmost column; I will now make an issue about naming of things in the interface so that we both have an overview of what we are talking about): 

However, I see a potential disaster on the horizon waiting for us (excuse me the long post, I had to bring down all my thoughts about this):
The segmentation of the bilingual texts and the monolingual quotations will not match up. 
So currently we have the case where a root segment a and a root segment b are pulled together in the root-entry (a+b). We might segment that and work with arrays to fix the issue as you suggested and as I also did before.  
Imagine now we have the Tibetan root text like it is right now:
segment a+b
segment c+d
segment e (longer segment more than 8 words)
segment f+g
Now the Sanskrit texts wants to match with these matches but it is asking for the Tibetan segments:
segment b+c
segment e+f
And so on. I am afraid that this could turn into a big problem if we want to provide more or less precise Sanskrit<>Tibetan matches. I am therefore inclined to do what I described in the earlier post, because if each segment in the root-text is independent, then the Sanskrit can ask for any combination of root segments. 
However a new problem arises: Imagine the Sanskrit (or Tibetan) quote/translation are much longer than the Tibetan segment (they might easily span multiple lines). In that case, we need to assign one quote to multiple root segments. 
A cheap and low-work solution for both of us is just to copy each quote as many times as we have root segments. So imagine we have a long_quote_b(3 segments), then the resulting csv table could look like this::
segment_a long_quote_b
segment_b long_quote_b
segment_c long_quote_b

This solution will make the frontend behave precisely as we want it to behave. The big issue here is however that it creates a lot of unnecessary overhead (all the copied quotes are just creating even more traffic and bring the webbrowser cose to crashing/OOM issues. 
One way to go about this: Mention each quote only once in the csv-file, but keep a list of the root-segments with the quote that indicates where it should match. Then we can generate a dictionary with all the quotes and iterate over the source segments on runtime and pull the quotes as we need them.
In order to make life easier for you, I can change the output of my neuronal networks so that it directly writes into JSON. I imagine it would make sense to have one dictionary for the root segments and one dictionary for the quotes. What do you think about this solution?

**Comment by sebastian-nehrdich** [comment]
Update: I added the JSON-version of the Tibetan CSV-files to the github (data-incoming/tibetan/json). A description is under data-incoming/tibetan/README.md. I could stripped a lot of (useless) information from the files and the size is now reduced drastically, almost by 50% compared to the CSV files.

**Comment by ayya-vimala** [comment]
I found a little bug in the last data-set.
T06TD4023E mentions 

>     {
>         "segmentnr": 310,
>         "segment": " mgon po byams-pas mdzad-pa rdzogs so ////",
>         "parallels": [
>             {
>                 "probability": 0.020919085,
>                 "parsegnr": [
>                     "T06TD4028E:529",
>                     "T06TD4028E:530"
>                 ],
>                 "parsegment": [
>                     "chos dang chos nyid rnam par 'byed pa'i 'grel pa slob dpon dbyig gnyen  gyis mdzad pa rdzogs so ////",
>                     "kha che'i pandi ta ma h'a dza na dang  lo ts'a ba blo ldan shes rab kyis bsgyur ba'o ////"
>                 ]
>             }
>         ]
>     },

But T06TD4028E:530 does not exist in T06TD4028E.

**Comment by sebastian-nehrdich** [comment]
thanks for pointing this out. Turns out that in each file, the last segment got lost because I didn't iterate including the last position. Will be fixed in the next upload!

**Comment by ayya-vimala** [comment]
Can you do me a big favor and replace all the '-' in Sanskrit filenames to '_'. That will save me a whole lot of issues.

**Comment by sebastian-nehrdich** [comment]
I will upload an updated version of the GRETIL data in a few hours including the changes we discussed via mail.
There will be some changes to the naming of the segments as opposed to the Tibetan JSONS.
There are two things we need to change:
1. the segments of the root-text (in the json this is 'segmentnr') will be changed from just a number to a string including the segment type seperated by a ':'. So the json ' "segmentnr": 1 ' will turn into: ' "segmentnr": "P:0_0" '. Also the numbering is different: P:0_0 means this segment is the first sentence in a  paragraph, "P:0_1" is the second sentence in a paragraph etc. In GRETIL, we have roughly two types of segments: P for paragraph and LG for line group. 
2. Regarding the segments in the quoted passages, we also need to change the naming scheme slightly. It is now going to be:  "srabhu_u:P:0_1", so it is filename:segmenttype:segmentnumber. 

I know these changes break the consistency with the Tibetan JSONs, but it's in my eyes necessary to be able to store all the desired information. So it might be good to change all the Tibetan JSON files as well (even though this will add some overhead to the JSON files). If you agree with that, I can do that and re-upload the Tibetan JSONs. 

I have tried a bit to reduce the file sizes of the Tibetan JSONs and the greatest reductions in filesize are probably achieved by using a list to store the quotes and calling them by an individual id for each segment (about 50% reduction in filesize) and by using lists without keys to store the data (another 30% reduction).
So with these two changes a 100mb file can be turned into a 30mb file, but the JSON might not be very human-readable after that. Removing self-quotations from the own file does not have a huge effect, somewhere in the range of 5-10% depending on the character of the file. We could also try a conservative hard limit on the max. number of quotes per segment. At the moment it is at 40, but we could just be strict and cut at 10, which will further reduce the filesize.

**Comment by ayya-vimala** [comment]
The fact that they are becoming string instead of number is not a problem. 
The change in JSON files will mean a lot of work for me to change the frontend and I'm not sure if I will be able to get that done this week.
With regards to number 2: can the naming scheme also include the directiory like "T06srabhu_u:P:0_1". Otherwise the frontend will constantly have to start looking for each individual parallels what directory goes with that.

I have to go so not much time to look into it right now. Maybe later ...

**Comment by sebastian-nehrdich** [comment]
Yes I agree that it is too much to change the layout of the JSON - we might be best of by aiming straight for a database. It seems unnecessary work for to me! Maybe we just accept the filesize at the moment.. 
Yes I am adding the folder-names to the filenames in the quotes, I forgot to mention in my post (sorry for this!)

**Comment by ayya-vimala** [comment]
I agree that going straight for a database might be much better. But will have to tell Orna which files to take for her presentation and mention that the site is not yet ready because of those large filesizes. Maybe I can add a warning on the homepage too.

Your idea to reduce the hard limit on the number of quotes to 10 might be a good idea to reduce some of the file sizes as well as reduce probability cutoff to 2 or so. I also noticed that in the Tibetan files the sanskrit parallels are not yet in it. So you have to redo them again anyway. 

I will start working to make a cutoff in paragraphs when the numbers start with 0.

**Comment by sebastian-nehrdich** [comment]
I have updated the Sanskrit files now on buddhist-db. Strangely, the Sanskrit jsons only are at 1,2gb fo the entire gretil collection now. To be fair, I filtered out segments longer than 3000 characters (which is a lot, actually) in order to avoid the problem with the long segments that we had yesterday. Anyway, this motivates me to investigate a bit deeper into what went wrong with the Tibean jsons. Something must have gone wrong, i am sure.

**Comment by ayya-vimala** [comment]
I had a look at the uploaded Sanskrit files but most of them are just 2B and just contain an empty array ... 
Also some seem to be missing altogether ... If anything went wrong it was with these files because at the very least they should contain some kind of segment text and number.

**Comment by ayya-vimala** [comment]
For instance, look at http://buddhist-db.de/vimala/suttas/sanskrit/GE07/
The first one (bhagvg_u.json) is missing. The second one does not have any parallels with the directory number in front on the name, the third one contains only rubbish parallels that are no parallels at all. 
Then most of them are just an empty array ....
The same with many of the others. 
I just took some samples here and there. The Tibetan Sanskrit onces seem to be a bit better, but there are also some really weird things. Here is one of 109 MB:
http://buddhist-db.de/vimala/suttas/sanskrit/T06/T06srabhu_u.json
But when you look at the data it does not make sense. There are arrays of parallels that do not have consequential numbers, the parallels don't make sense and don't match the same numbers as parallel segments ...

**Comment by ayya-vimala** [comment]
Also other things with the numbering. For instance in http://buddhist-db.de/vimala/suttas/sanskrit/T06/T06bsa010_u.json there are parallel number like: **T06vmvkbh_u:LG:Mvk_1.2_0** but that number does not appear in **T06vmvkbh_u:LG**.
It seems that there are masses of issues so something must have gone wrong .... :-(

**Comment by sebastian-nehrdich** [comment]
Thank you for pointing this out!
I decided to go back to the original line breaks of the GRETIL html files which should solve the problems that you and I have observed with these files in the last days. I was kind of over-optimistic by assuming that we can use regex +  occurrences of dots, slashes and dandas as markers for sentences but this seems to give unreliable results and is breaking the formatting of the json files. The original GRETIL line breaks aren't great for segmentation either because they frequently lie in the middle of sentences etc. but at least they should not break our output files. Anyway, I think to do this properly we will need good XML of the rest of GRETIL and that's far from done yet (but might come at some point!). 
Parsing the GRETIL HTML with regex feels like a temporary stay in the avīci hell and I now understand very well why we are not supposed to do that. One might find a solution for a certain subset of texts but for sure, this is going to break compatibility with another subset.

**Comment by ayya-vimala** [comment]
![avici01](https://user-images.githubusercontent.com/49606945/60487714-5a962300-9ca0-11e9-8804-bef41990f5d6.jpg)
Image of working with Gretil files ... aaaahhhh!
Just to remind you of our planned meeting this afternoon at 12 CET.

**Comment by sebastian-nehrdich** [comment]
The updated Tibetan files are now uploading Sanskrit will follow soon.
I changed two things:
1. replaced the similarity value with an int value between 0 and 100 (100 = perfect similarity)
2. cut-off at max 10 parallels per segment

With these measures, the size of the entire collection was reduced from 16GB to just 6GB!

**Comment by ayya-vimala** [comment]
Hold on ... so this has changed i.e. before 0 was the highest similarity and the higher was less similarity right? Now that's swapped around? Then I need to adjust the filters to swap too.

Great that we have less GBs :-) I noticed not all Tibetan files have changed yet.

**Comment by sebastian-nehrdich** [comment]
Yes I find it somewhat more intuitive to say 'we have 100% similarity', otherwise we would have to say '100% divergence' which is also not the end of the world but I guess that the users might be puzzled by that! 
Yepp RSYNC is still running, I don't know how long it will take, maybe another hour or two until everything is updated.

**Comment by ayya-vimala** [comment]
Yes, you are right. It is more logical. Sure, will change it.

**Comment by ayya-vimala** [comment]
So explain to me these scores: They look like a perfect match to me. Or is that because it takes into account the sentences before and after too and if they are really not similar, the score is 0.

>     {
>         "segmentnr": 18,
>         "segment": " chos kyang don rnams phye bas mkhas rnams 'di na dga' mchog skyed //",
>         "parallels": [
>             {
>                 "probability": 19,
>                 "parsegnr": [
>                     "T06TD4026E:58"
>                 ],
>                 "parsegment": [
>                     "chos kyang don rnams phye bas mkhas rnams 'di na dga' mchog bskyed //"
>                 ]
>             },
>             {
>                 "probability": 0,
>                 "parsegnr": [
>                     "T03TD3796E:296"
>                 ],
>                 "parsegment": [
>                     "chos kyang don rnams phye bas mkhas rnams 'di na dga' mchog skyed //"
>                 ]
>             }
>         ]
>     },

**Comment by sebastian-nehrdich** [comment]
You are right, I forgot that we are no aiming for the highest possible score and not the lowest, so this is a bug in the file-generation. I will upload the fixed files today!

**Comment by sebastian-nehrdich** [comment]
The Tibetan files are now updated and the bug you mentioned in your last posting is removed.
As far as I can judge, the Tibetan data does not seem to show any major problems anymore (but please correct me if I am wrong). One thing that is still not perfect is the fact that sometimes quotes are also assigned to segments before or after the quote which do not show that much similarity. I have to rework the matching algorithm in order to be more precise in such cases, it is certainly possible to solve this! 

The Sanskrit data is also updated now. The major bugs I fixed:
1. No more empty JSON files
2. Segment numbers now should all be correct (let's hope!)
3. Solved the problem with the super-long parallels

Some problems remain: 
1. Oddly formatted segments that are sometimes very long, sometimes very short, sometimes include some kind of markup or sometimes don't include any markup - almost impossible to solve without proper XML formatting
2. If a parallels spans over LG and and P-type of segments, the ordering of the segments is messed up. To make clear; the ordering of the text is still ok, it is just the case that the order of the segments given in the parallel is messed up. For example: 
`{
                "probability": 100,
                "parsegnr": [
                    "XXbst-108u:LG:BuSto_64.1_0",
                    "XXbst-108u:LG:BuSto_64.1_1",
                    "XXbst-108u:P:338_0"
                ],
                "parsegment": [
                    "om namaḥ śrīprajñāpāramitāyai /",
                    "nirvikalpe namastubhyaṃ prajñāpāramite 'mite /",
                    "yā tvaṃ sarvānavadyāṅgi niravadyairnirīkṣyase //"
                ]
            },
`
It should be the other way round, e.g. first the P: and after that the LG. This happens at the moment because python's sort-function is alphanumerical which works fine in the case that we only have the same type of segments but breaks when they are mixed. Anyway, I will think about a way to preserve order in the parallels without  using the sort-function and update that at some point. 

Also the bilingual data is now added to the jsons in both languages and I am currently replacing the relevant JSONS on the server (will be done in a few minutes). I will open a seperate issue for this.

**Comment by ayya-vimala** [comment]
Will have a look. What is the addition of _added to T064020E_added.json_added?

**Comment by sebastian-nehrdich** [comment]
I am just updating now because I noticed that the similarity value was changed to the new format, will be changed in a few minutes. 
The _added files don't have anything to say, I will delete them!

**Comment by ayya-vimala** [comment]
In the sanskrit files, the following is not consistent: (example from T06bsa010_u.json)
```
            {
                "probability": 100,
                "parsegnr": [
                    "T06vmvkbh_u:P:6_0",
                    "T06vmvkbh_u:P:6_1"
                ],
                "parsegment": [
                    "<b>abhūtaparikalpo 'sti dvayan tatra na vidyate</b> /",
                    "<b></b>  <b>śūnyatā vidyate tv atra tasyām api sa vidyate</b> /"
                ]
            },
            {
                "probability": 92,
                "parsegnr": [
                    "GR14vdhasrtu:P13421_0",
                    "GR14vdhasrtu:P13422_0",
                    "GR14vdhasrtu:P13423_0",
                    "GR14vdhasrtu:P13424_0"
                ],
                "parsegment": [
                    "vidyate kurusattama__Vdha_036.009",
                    "vidyate nātra kaścana__Vdha_038.052",
                    "vidyate madhusūdana__Vdha_044.003",
                    "vidyate hy atra janmani__Vdha_030.002"
                ]
            },
```
So for the first set of parallels, the numbers are in the form of P:6_0. For the second set they are in the form of P13421_0. (so without the `:` behind the P). I don't mind whichever one it is, but it will be good to be consistent with these things.

**Comment by ayya-vimala** [comment]
Another strange thing I found in the same file: segment number P:0_0 appears twice; once at the top and once at the bottom.
top:
```
    {
        "segmentnr": "P:0_0",
        "segment": " Āryamaitreyapraṇītā Madhyāntavibhāgakārikā (Mvk) /",
        "parallels": [
            {
                "parsegment": [
                    "//"
                ],
                "parsegnr": [
                    "TD4021E:0"
                ],
                "probability": 100
            }
        ]
    },
```
bottom:
 ```
   {
        "segmentnr": "P:0_0",
        "segment": "Ārya Saṃghāṭa Sūtra /",
        "parallels": [
            {
                "parsegment": [
                    "//"
                ],
                "parsegnr": [
                    "TD4021E:0"
                ],
                "probability": 100
            },
            {
                "probability": 60,
                "parsegnr": [
                    "XXbsu041_u:P10878_0",
                    "XXbsu041_u:P10879_0"
                ],
                "parsegment": [
                    "mahābodhimantrā sarvakarmika sarvārthasampādikāḥ sarvānarthapratighātikāḥ",
                    "sarvabuddhadharmāṃ pāripūrikā sarvakleśāṃ niṣūdinī sarvamantrāṃ"
                ]
            }
        ]
    }
```

**Comment by ayya-vimala** [comment]
Another inconsistencey with paragraph numbers: In T06bsa010_u.json, the first paragraph is marked as:
```
        "segmentnr": "P:1_0",
        "segment": " Lakṣaṇaparicchedaḥ prathamaḥ /",
```
But then it is referred to it in T06TD4021E.json as:
```
                "parsegment": [
                    "Lakṣaṇaparicchedaḥ prathamaḥ /"
                ],
                "parsegnr": [
                    "T06bsa010_u:P#1_0"
                ],
```
So this cannot find the right segment number because it is looking for P#1_0 and not for P:1_0.

**Comment by sebastian-nehrdich** [comment]
Regarding the problems with the paragraph names: I have fixed these now and it will be ok in the next upload of the files. 
Regarding the issue with the strange segments showing up at the end of the file: This is actually the first segment of the next file (imagine we are converting them into a big stream in order to be able to calculate the possible parallels. This is not trivial to debug and I suggest that we bear with it for the time being until I find some time in the next weeks to fix it. 
I have also discovered a larger issue regarding malicious tab-characters in the input files that break the alignment process and create nonsense-matches (can be seen very well in the sanskrit version of the abhidharma-kośa-bhāṣya). I could fix it but I also have to regenerate the data because of that!

**Comment by ayya-vimala** [comment]
Every time you reply I wonder what time it is in Tokyo!
I know how annoying this is because every time you change something it takes hours to regenerate the files and upload them again!
If the bug with the last P:0_0 to show up at the end of each one is consistent, I can just get the system to discard the last entry.

**Comment by sebastian-nehrdich** [comment]
Don't worry about the time, thanks to Harunaga I  already got used to
living in two timezones at the same time! :)
Yes it's annoying but we have to go through with these issues in order to
get a proper result, so nevermind! The end is slowly getting in sight.
Regarding the last paragraph-problem, I am afraid it doesn't seem to be
consistent. I checked a couple of other files and it is not always like
that. We could try to ignore a P:0_0 at the end of the file, I think that
should work as a temporary fix. But this will break again if the next file
is starting with a LG-segment, so... Let's hope that case never occours
(but I am afraid it does).

**Comment by sebastian-nehrdich** [comment]
Small update: I have now fixed the problem with the P:0_0 occouring at the end of the files, so you don't need to worry about this. Should not be there in the next generation of Sanskrit files. However this fix is not able to deal with the case that an LG appears wrongly at the end of the line so we need to keep our eyes open for that case.

**Comment by sebastian-nehrdich** [comment]
Another day, a new attempt: I have regenerated the entire dataset (Tibetan/Sanskrit/Bilingual) in order to remove the mistakes that we discovered yesterday.
The new Tibetan data together with the Sanskrit parallels is already online.
Sanskrit is uploading now but it is not yet done (maybe another hour or two).
This are the currently available bilingual Tibetan files: 
 tibetan/T01/T01TD1127E.json
tibetan/T06/T06TD4036E.json
tibetan/K12/K12kl00557.json
tibetan/T02/T02TD2531E.json
tibetan/T10/T10TD4183E.json
tibetan/T11/T11TD4264E.json
tibetan/T04/T04TD3825E.json
tibetan/T04/T04TD3861E.json
tibetan/T04/T04TD3958E.json
tibetan/T04/T04TD3955E.json
tibetan/T06/T06TD4058E.json
tibetan/T04/T04TD3828E.json
tibetan/T11/T11TD4213E.json
tibetan/T10/T10TD4158E.json
tibetan/T11/T11TD4213E.json
tibetan/T02/T02TD3721E.json
tibetan/T02/T02TD1748E.json
tibetan/T11/T11TD4206E.json
tibetan/T04/T04TD3947E.json
tibetan/T11/T11TD4238E.json
tibetan/T04/T04TD3977E.json
tibetan/T03/T03TD3792E.json
tibetan/T04/T04TD3827E.json
tibetan/T04/T04TD3902E.json
tibetan/T06/T06TD4020E.json
tibetan/T02/T02TD1798E.json
tibetan/T04/T04TD3828E.json
tibetan/T10/T10TD4158E.json
tibetan/T01/T01TD1127E.json
tibetan/T02/T02TD3496E.json
tibetan/T06/T06TD4037E.json
tibetan/T02/T02TD2693E.json
tibetan/T06/T06TD4049E.json
tibetan/T02/T02TD1353E.json
tibetan/T10/T10TD4182E.json
tibetan/T06/T06TD4036E.json
tibetan/T01/T01TD1128E.json
tibetan/T03/T03TD3786E.json
tibetan/T09/T09TD4152E.json
tibetan/T06/T06TD4074E.json
tibetan/T02/T02TD2460E.json
tibetan/T11/T11TD4218E.json
tibetan/T04/T04TD3845E.json
tibetan/T03/T03TD3817E.json
tibetan/K10/K10kl00095e.json
tibetan/T09/T09TD4152E.json
tibetan/T04/T04TD3887E.json
tibetan/T02/T02TD3125E.json
tibetan/T03/T03TD3797E.json
tibetan/T06/T06TD4026E.json
tibetan/T11/T11TD4214E.json
tibetan/T02/T02TD1748E.json
tibetan/T11/T11TD4218E.json
tibetan/T02/T02TD3141E.json
tibetan/T06/T06TD4020E.json
tibetan/T01/T01TD1120E.json
tibetan/K10/K10kl00107e.json
tibetan/K07/K07kl00014einc.json
tibetan/T11/T11TD4210E.json
tibetan/T06/T06TD4059E.json
tibetan/T06/T06TD4021E.json
tibetan/T04/T04TD3841E.json
tibetan/T02/T02TD3125E.json
tibetan/T04/T04TD3846E.json
tibetan/T10/T10TD4162E.json
tibetan/T02/T02TD1802E.json
tibetan/T07/T07TD4101E.json
tibetan/T11/T11TD4235E.json
tibetan/T06/T06TD4055E.json
tibetan/T04/T04TD3973E.json
tibetan/K10/K10kl00102e.json
tibetan/T02/T02TD3495E.json
tibetan/T04/T04TD3825E.json
tibetan/T04/T04TD3860E.json
tibetan/T03/T03TD3786E.json
tibetan/T04/T04TD3944E.json
tibetan/T05/T05TD3986E.json
tibetan/T11/T11TD4211E.json
tibetan/T04/T04TD3957E.json
tibetan/T11/T11TD4212E.json
tibetan/T04/T04TD3969E.json
tibetan/T11/T11TD4261E.json
tibetan/T04/T04TD3836E.json
tibetan/T11/T11TD4238E.json
tibetan/T04/T04TD3953E.json
tibetan/K12/K12kl00361.json
tibetan/T11/T11TD4212E.json
tibetan/T02/T02TD3504E.json
tibetan/T11/T11TD4261E.json
tibetan/T02/T02TD1804E.json
tibetan/T04/T04TD3832E.json
tibetan/T11/T11TD4205E.json
tibetan/T10/T10TD4158E.json

**Comment by ayya-vimala** [comment]
Thanks so much. That looks great. The only thing I still see right now is that in the Sanskrit files, the parallels to the Tibetan texts do not have the collection number in front. So I get a link to for instance TD4026E and it cannot find that because it does not know where to look. The other way around is fine.

**Comment by ayya-vimala** [comment]
Actually, I also found the issue here: http://buddhist-db.de/vimala/suttas/sanskrit/XX/XXmaricdhu.json
where it says: `"nagsu_pu:P:2_1"` but that sort of thing can also have to do with the crappy Gretil files.

**Comment by sebastian-nehrdich** [comment]
Thank you for pointing out the problems! I am now fixing the issue with the
missing categories in the Tibetan translations, that's easily solvable.
Regarding GRETIL / nagsu_pu: There is a small number of GRETIL files (maybe
20 or so altogether) where the category linking didn't work for some
reason. I don't clearly remember why it didn't work out but I suggest to
live with that until we get proper XML of GRETIL and have to re-do this
anyway. It's not such a big problem atm I think...

**Comment by sebastian-nehrdich** [comment]
New data is online now.

**Comment by ayya-vimala** [comment]
Looks good. Thanks!

**Comment by sebastian-nehrdich** [comment]
Some strange bug on earth! Each Sanskrit JSON actually has repeated it's content four times!!!!
17GB of Sanskrit is indeed strange. ^^
I will fix that today in the afternoon. 
Another thing: As far as I can see, the scrolling of the parallel text on the right side when it is Sanskrit does not yet seem to work, it opens the file but it's not getting scrolled properly, am I right? 
I also remember something else not working, but I will give you a more detailed description later.

**Comment by ayya-vimala** [comment]
I cannot see this bug. It scrolls fine for me. Can you file a bug-report with the exact steps? You can do that by making a new issue and then at the top it says `Template: none` - change that to `bug-report` and then just complete the various parts where needed, also possible screenshots. Also important to see which files you are opening because there are many many choices!!!

**Comment by sebastian-nehrdich** [comment]
OK I filled in a bug report - I hope it is not my faulty data causing this
problem!
I am now regenerating the Sanskrit data (without the 4x repetition ;)).
